### ISSUE TITLE
Lazily load default value of environment variable

### ISSUE THREAD

**Issue by konoufo** [joke2k/django-environ#185]
Would you accept a pull request to enable lazy evaluation of default value ? I think there are several use cases where you want to generate the default on-demand. For example, creating a new temporary directory and use it as a default value when a env var doesn't exist in environment.

**Comment by thomasf** [comment]
Please describe how you suggest it will work!

I use a pattern like this when I want to have that feature and it's not very verbose:

```py

env = environ.Env(
    SOMETHING=(str),
)

SOMETHING = env("SOMETHING",default=None)
if not SOMETHING:
    ... calculate and assign value of something
```

To me this has worked fine..

**Comment by konoufo** [comment]
Yeah I think being able to pass a plain function that's evaluated when `env("SOMETHING", fun)` is called. This wouldn't make a difference timing-wise compared to what we're doing right now.

**Comment by thomasf** [comment]
So something like this then..

```py

def create_something():
    return "SOMETHING"

env = environ.Env(
    SOMETHING=(str, create_something),
)

SOMETHING = env('SOMETHING')

OTHER_THING=env('OTHER_THING', default_value=create_something, cast_to=str)

```

**Comment by konoufo** [comment]
Yeah this is how I'd use it. I could submit a PR if you're okay with this.

________________________________
De : Thomas Frössman <notifications@github.com>
Envoyé : dimanche 1 juillet 2018 11:02:16
À : joke2k/django-environ
Cc : konoufo; Author
Objet : Re: [joke2k/django-environ] Lazily load default value of environment variable (#185)


So something like this then..

def create_something():
    return "SOMETHING"

env = environ.Env(
    SOMETHING=(str, create_something),
)

SOMETHING = env('SOMETHING')

OTHER_THING=env('OTHER_THING', default_value=create_something, cast_to=str)

**Comment by thomasf** [comment]
I am not a maintainer. I just wanted to understand what you were requesting, it looks clear to me now..

**Comment by konoufo** [comment]
Oh okay! Sorry. Alright, waiting for the maintainers then...

________________________________
De : Thomas Frössman <notifications@github.com>
Envoyé : dimanche 1 juillet 2018 17:45:57
À : joke2k/django-environ
Cc : konoufo; Author
Objet : Re: [joke2k/django-environ] Lazily load default value of environment variable (#185)


I am not a maintainer. I just wanted to understand what you were requesting, it looks clear to me now..

**Comment by Jafnee** [comment]
I haven't tested this, but would creating your own lazy class work?

```python
def create_something():
    return "SOMETHING"

class LazyString:
    def __init__(self, value, fn):
        self.value = value
        self.fn = fn

    def __str__(self):
        if self.value:
            return self.value
        return self.fn()


OTHER_THING=LazyString(env('OTHERTHING', default=None), create_something)
```

**Comment by konoufo** [comment]
Well, I'd have to edit every code that uses `OTHER_THING` to replace `settings.OTHER_THING` with `settings.OTHER_THING.value`. I'm aware of workarounds to get the functionality I want. But for the sake of separation of concerns, I don't want to call `create_something` inside the settings file. And for consistency, I want the actual value of a setting to be always directly accessible thru `settings.FOO`. Moreover, I think to have the `default` argument to also accept a function just make sense and this pattern is ubiquitous in Python libraries.

________________________________
De : Jafnee <notifications@github.com>
Envoyé : lundi 2 juillet 2018 03:29:24
À : joke2k/django-environ
Cc : konoufo; Author
Objet : Re: [joke2k/django-environ] Lazily load default value of environment variable (#185)


I haven't tested this, but would creating your own lazy class work?

def create_something():
    return "SOMETHING"

class LazyString:
    def __init__(self, value, fn):
        self.value = value
        self.fn = fn

    def __str__(self):
        if self.value:
            return self.value
        return self.fn()


OTHER_THING=LazyString(env('OTHERTHING', default=None), create_something)

**Comment by Jafnee** [comment]
Shouldn't need to replace everything to `settings.OTHER_THING.value`, the example above is meant to evaluated lazily whenever the setting is used.
The value will either be the string set via env var or the return value of the function provided. The function is also called each time the setting is used, e.g if it returned the timestamp now, it will always return the latest value.

**Comment by thomasf** [comment]
Remember that if you are using it for creating a new temporary directory and use it as a default value and don't return a static value the value will probably be different for each django process in a uwsgi server. If possible you should normally always prefer a deterministic way to create the settings.

**Comment by joke2k** [comment]
Hi, I think this feature could be useful.
As @konoufo said, callable defaults are ubiquitous in Python, and we could implement it as easy.
Here https://github.com/joke2k/django-environ/blob/develop/environ/environ.py#L279 we could handle this, without side effects.

**Comment by konoufo** [comment]
@Jafnee  that is if you assume that the setting will be used in some way as a string or coerced to a string.
@thomasf I realize that and the temporary directory setting is just a specific use case; but yeah I'm aware that it has to be deterministic if I want a single unique directory for every process.
@joke2k Right, I could submit a PR.
